src/cpuid.c

```diff
@@ -311,7 +311,7 @@ const char *cpu_get_model (void)
 
 const char *cpu_get_freq_info1 (void)
 {
-  static char  result [100];
+  static char  result [102];
```

Comment:
static in this context allocates memory as if it's global value (100-102 bytes for the duration of the program) but is only accessible to this function when it next runs. Is this an error?

Reply:
>  Is this an error?

`cpu_get_freq_info1()` is not called. Perhaps better to have it fill a user-supplied buffer;
```c
const char *cpu_get_freq_info1 (char *ret)
{
  ...
}
```
And same for `cpu_get_freq_info2 (char *ret)`.

Reply:
That's odd. The only call to either `cpu_get_freq_info` function I can find is within a `#if 0` statement which itself is in a conditional main function:
https://github.com/gvanem/Watt-32/blob/631655e45083955df0ec5abf89b37c6688e24237/src/cpuid.c#L373-L401
I checked with both [GitHub search](https://github.com/search?q=repo%3Agvanem%2FWatt-32%20cpu_get_freq_info&type=code) and `grep` on my local copy
```
$ grep cpu_get_freq_info Watt-32/ -R
grep: Watt-32/src/BUILD/DJGPP/PCDBUG.O: binary file matches
Watt-32/src/cpuid.c:const char *cpu_get_freq_info1 (void)
Watt-32/src/cpuid.c:const char *cpu_get_freq_info2 (void)
Watt-32/src/cpuid.c:  freq = cpu_get_freq_info1();
Watt-32/src/cpuid.c:  freq = cpu_get_freq_info2();
grep: Watt-32/lib/LIBWATT.A: binary file matches
grep: Watt-32/lib/WATTCPWF.LIB: binary file matches
```
Am I missing something? Should we remove these `cpu_get_freq_info` functions completely?

Reply:
246be9c wraps these functions in a pre-processor condition and frees 200 bytes of global memory allocation for any linker that wasn't smart enough to strip these functions automatically

Reply:
> Am I missing something?

No. Do not think so.

>  for any linker that wasn't smart enough to strip these functions automatically

And are some linkers (which?) smart enough to strip-off these `static char` buffers too?

Reply:
> And are some linkers (which?) smart enough to strip-off these `static char` buffers too?

You would hope so but the only way to know for certain what any particular C compiler/linker will do is to compile then disassemble the binary. [Compiler Explorer](https://github.com/compiler-explorer/compiler-explorer) and [Ghidra](https://github.com/NationalSecurityAgency/ghidra) are two useful tools for this.

Take this [example](https://godbolt.org/z/K6f9Trc94):
Modern `gcc` decides to keep `const char *uselessFunction(char *ptr)` even though it's unused. GCC has no way of knowing if its expected by something else on a Linux target even though it's a binary another program might link it and use `uselessFunction()`. 
```c
/* Some other program */
#include <dlfcn.h>
#include <stdio.h>

int main(int argc, char **argv) {
   void *bin = dlopen("a.out", RTLD_LAZY);
   const char *(*func)(char *) = dlsym(bin, "uselessFunction");
   const char *str = (*func)(argv[0]);
   puts(str);
   dlclose(bin);
   return 0;
}
```

If you write `static const char *uselessFunction(char *ptr)` then gcc knows for certain that this function will never be called outside of this source file and will strip it out unless `-O0` is used. In contrast a DOS focused compiler doesn't have to worry about linking and symbols from external sources and might always strip unused functions.

It's also worth noting how `gcc` replaces `sprintf(x, "%s", y)` with `strcpy(x, y)`. The only rule all C compiler optimizations must obey is that the output must be guaranteed to be identical to the codes expectation.